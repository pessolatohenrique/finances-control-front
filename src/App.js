import React from "react";
import AppBarWrapper from "./components/AppBarWrapper";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import ExampleList from "./views/register/ExampleList";
import ExampleForm from "./views/register/ExampleForm";
import ExampleDetail from "./views/register/ExampleDetail";
import NotFound from "./components/NotFound";
import LoginForm from "./views/register/LoginForm";

function App() {
  return (
    <Router>
      <AppBarWrapper />

      <Switch>
        <Route exact path="/signin">
          <LoginForm />
        </Route>

        <Route exact path="/livros">
          <ExampleList />
        </Route>

        <Route exact path="/autores/novo">
          <ExampleForm />
        </Route>

        <Route exact path="/livros/:id">
          <ExampleDetail />
        </Route>

        <Route path="*">
          <NotFound />
        </Route>
      </Switch>
    </Router>
  );
}

export default App;
