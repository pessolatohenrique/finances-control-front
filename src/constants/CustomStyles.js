import { styled } from "@mui/material/styles";
import { tableCellClasses } from "@mui/material/TableCell";
import { TableCell } from "@mui/material";

export const StyledTableCell = styled(TableCell)(({ theme }) => ({
  [`&.${tableCellClasses.head}`]: {
    fontSize: "1.1em",
    fontWeight: 300,
  },
}));
