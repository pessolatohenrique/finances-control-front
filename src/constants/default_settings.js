export const PROJECT_NAME = "Project Name";
export const THEME_COLOR = "#1976d2";
export const SNACKBAR_DIRECTION = { vertical: "top", horizontal: "right" };
