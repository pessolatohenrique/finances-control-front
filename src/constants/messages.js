export const REQUIRED_MESSAGE = "Campo de preenchimento obrigat√≥rio";
export const USER_NOT_EXISTS_MESSAGE = "This user doesnt exists";
export const WRONG_PASSWORD_MESSAGE = "Invalid username or password";
