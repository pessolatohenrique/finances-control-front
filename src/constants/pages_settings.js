export const PATHS_WITHOUT_LOGIN = ["/usuario/novo"];
