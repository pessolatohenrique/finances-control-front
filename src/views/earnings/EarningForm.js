import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import {
  Container,
  Card,
  CardContent,
  Typography,
  Grid,
  TextField,
  Button,
  Snackbar,
  Alert,
} from "@mui/material";
import { AdapterDateFns } from "@mui/x-date-pickers/AdapterDateFns";
import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
import { DatePicker } from "@mui/x-date-pickers/DatePicker";

import axios from "axios";
import moment from "moment";
import NumberFormat from "react-number-format";
import { Controller, useForm } from "react-hook-form";

import {
  THEME_COLOR,
  SNACKBAR_DIRECTION,
} from "../../constants/default_settings";
import { REQUIRED_MESSAGE } from "../../constants/messages";
import { formatMoneyToDecimal } from "../../utils/formatters";
import useToast from "../../hooks/useToast";
import BreadcrumbsWrapper from "../../components/BreadcrumbsWrapper";

const RESET_FIELDS = { name: "", value: "", transaction_date: new Date() };

function EarningForm() {
  const {
    control,
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm({
    defaultValues: RESET_FIELDS,
  });

  const { id } = useParams();

  useEffect(() => {
    async function loadData() {
      try {
        const response = await axios.get(`/user_earning/${id}`);
        return response?.data;
      } catch (error) {}
    }

    async function verifyUpdate() {
      if (id) {
        const data = await loadData();
        const transaction_date = moment(data.transaction_date, "YYYY-MM-DD");

        reset({
          name: data?.Earning.name,
          value: data?.value,
          transaction_date,
        });
      }
    }

    verifyUpdate();
  }, [id, reset]);

  const {
    open,
    error,
    messageType,
    setError,
    showToast,
    hideToast,
  } = useToast();

  const formatDataSubmit = (data) => {
    const { transaction_date, value } = data;

    if (id) {
      return {
        ...data,
        value: parseFloat(formatMoneyToDecimal(value)),
        transaction_date: moment(transaction_date, "DD/MM/YYYY").format(
          "YYYY-MM-DD"
        ),
      };
    }

    return {
      earnings: [
        {
          ...data,
          transaction_date: moment(transaction_date, "DD/MM/YYYY").format(
            "YYYY-MM-DD"
          ),
          isPublic: 1,
          value: formatMoneyToDecimal(value),
        },
      ],
    };
  };

  const verifyOperation = async (dataSubmit) => {
    if (id) {
      await axios.put(`/user_earning/${id}`, { ...dataSubmit });
      return;
    }

    await axios.post(`/user_earning`, { ...dataSubmit });
    reset(RESET_FIELDS);
  };

  const onSubmit = async (data) => {
    try {
      const dataSubmit = formatDataSubmit(data);
      await verifyOperation(dataSubmit);
      showToast("success");
      setError("Ganho salvo com sucesso!");
    } catch (error) {
      showToast();
      setError(error?.response?.data);
    }
  };

  return (
    <>
      <Container fixed>
        <Snackbar
          anchorOrigin={SNACKBAR_DIRECTION}
          open={open}
          autoHideDuration={6000}
          data-testid="snackbar"
          onClose={hideToast}
        >
          <Alert
            onClose={hideToast}
            severity={messageType}
            sx={{ width: "100%" }}
          >
            {error}
          </Alert>
        </Snackbar>
        <br />
        <Card>
          <CardContent>
            <BreadcrumbsWrapper
              parentLink={{ link: "/", label: "Dashboard" }}
              childrenLabel={id ? "Editar" : "Novo"}
            />
            <Typography
              gutterBottom
              variant="h5"
              component="h1"
              color={THEME_COLOR}
            >
              {id ? "Editar" : "Novo"} Ganho
            </Typography>
            <form onSubmit={handleSubmit(onSubmit)}>
              <Grid container direction={"column"} spacing={2}>
                <Grid item>
                  <TextField
                    fullWidth
                    id="name"
                    label="Descrição"
                    inputProps={{ "data-testid": "name" }}
                    error={Boolean(errors.name)}
                    helperText={errors.name && REQUIRED_MESSAGE}
                    {...register("name", {
                      required: true,
                    })}
                  />
                </Grid>
                <Grid item>
                  <Controller
                    render={({
                      field: { onChange, value },
                      fieldState: { error },
                    }) => (
                      <NumberFormat
                        thousandSeparator={"."}
                        decimalSeparator={","}
                        prefix={"R$ "}
                        customInput={TextField}
                        inputProps={{ "data-testid": "value" }}
                        fullWidth
                        label="Valor"
                        onChange={onChange}
                        value={value}
                        error={Boolean(error)}
                        helperText={Boolean(error) && REQUIRED_MESSAGE}
                      />
                    )}
                    name="value"
                    control={control}
                    rules={{ required: true }}
                  />
                </Grid>
                <Grid item>
                  <Controller
                    render={({
                      field: { onChange, value },
                      fieldState: { error },
                    }) => (
                      <LocalizationProvider dateAdapter={AdapterDateFns}>
                        <DatePicker
                          label="Data da transação"
                          inputProps={{ "data-testid": "transaction_date" }}
                          inputFormat="dd/MM/yyyy"
                          value={value}
                          onChange={onChange}
                          error={Boolean(error)}
                          helperText={Boolean(error) && REQUIRED_MESSAGE}
                          renderInput={(params) => (
                            <TextField
                              inputProps={{ "data-testid": "transaction_date" }}
                              fullWidth
                              {...params}
                            />
                          )}
                        />
                      </LocalizationProvider>
                    )}
                    name="transaction_date"
                    control={control}
                    rules={{ required: true }}
                  />
                </Grid>

                <Grid item>
                  <Button
                    variant="contained"
                    type="submit"
                    data-testid="submit-button"
                  >
                    Salvar
                  </Button>
                </Grid>
              </Grid>
            </form>
          </CardContent>
        </Card>
      </Container>
    </>
  );
}
export default EarningForm;
