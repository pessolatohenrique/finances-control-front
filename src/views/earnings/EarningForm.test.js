import {
  render,
  screen,
  fireEvent,
  waitForElement,
  waitForDomChange,
  waitFor,
} from "@testing-library/react";
import { REQUIRED_MESSAGE } from "../../constants/messages";
import App from "../../App";
import EarningForm from "./EarningForm";

describe("Dashboard Container", () => {
  beforeAll(() => {
    render(<App />);
    const submitBtn = screen.getByTestId("submit-button");
    const username = screen.getByTestId("username");
    const password = screen.getByTestId("password");

    fireEvent.input(username, {
      target: { value: process.env.REACT_APP_API_DEFAULT_EMAIL },
    });
    fireEvent.input(password, {
      target: {
        value: process.env.REACT_APP_API_DEFAULT_PASSWORD,
      },
    });
    fireEvent.click(submitBtn);
  });

  it("should generate snapshot", () => {
    const { container } = render(<EarningForm />);
    expect(container).toMatchSnapshot();
  });

  it("should not send form when field is empty", async () => {
    render(<EarningForm />);
    const submitBtn = screen.getByTestId("submit-button");
    const name = screen.getByTestId("name");

    fireEvent.input(name, {
      target: { value: "Teste descrição" },
    });

    fireEvent.click(submitBtn);

    await waitForElement(() => screen.getAllByText(REQUIRED_MESSAGE));
  });

  it("should send form when field is fullfiled", async () => {
    render(<EarningForm />);
    const submitBtn = screen.getByTestId("submit-button");
    const name = screen.getByTestId("name");
    const value = screen.getByTestId("value");

    fireEvent.input(name, {
      target: { value: "Teste descrição" },
    });

    fireEvent.input(value, {
      target: { value: "1500" },
    });

    fireEvent.click(submitBtn);

    await waitFor(() => expect(screen.getByTestId("snackbar")));
  });
});
