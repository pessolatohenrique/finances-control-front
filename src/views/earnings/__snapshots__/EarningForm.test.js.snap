// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Container should generate snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg MuiContainer-fixed css-1ss2ei0-MuiContainer-root"
  >
    <br />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <nav
          aria-label="breadcrumb"
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          >
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-9smqfy-MuiTypography-root-MuiLink-root"
                data-testid="bread-firstlevel"
                href="/"
              >
                Início
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              ›
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-9smqfy-MuiTypography-root-MuiLink-root"
                data-testid="bread-secondlevel"
                href="/"
              >
                Dashboard
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              ›
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
                data-testid="bread-lastlevel"
              >
                Novo
              </p>
            </li>
          </ol>
        </nav>
        <h1
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1y15g74-MuiTypography-root"
        >
          Novo Ganho
        </h1>
        <form>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column css-1rl1og4-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="name"
                  id="name-label"
                >
                  Descrição
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    data-testid="name"
                    id="name"
                    name="name"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-1ftyaf0"
                    >
                      <span>
                        Descrição
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                inputmode="numeric"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="mui-1"
                  id="mui-1-label"
                >
                  Valor
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    data-testid="value"
                    id="mui-1"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-1ftyaf0"
                    >
                      <span>
                        Valor
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for="mui-2"
                  id="mui-2-label"
                >
                  Data da transação
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    aria-label="Choose date, selected date is May 20, 2022"
                    aria-readonly="true"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id="mui-2"
                    readonly=""
                    type="text"
                    value="20/05/2022"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-1in441m"
                    >
                      <span>
                        Data da transação
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            >
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
                data-testid="submit-button"
                tabindex="0"
                type="submit"
              >
                Salvar
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
